File: 02/python/main.py
from typing import Iterator

from python_utils.input import add_argparse, advent_input

get_int = lambda x: int(x.strip())

action_funcs = {
    "up": lambda x: -x,
    "down": lambda x: x,
}

@advent_input("02/input_1.txt")
def task_one(lines: Iterator[str]) -> None:
    depth = 0
    horizontal = 0

    for line in lines:
        action, value = line.strip().split(" ")
        value = get_int(value)

        if action in action_funcs:
            depth += action_funcs[action](value)
        else:
            horizontal += value

    print(depth * horizontal)


@advent_input("02/input_1.txt")
def task_two(lines: Iterator[str]) -> None:
    depth = 0
    horizontal = 0
    aim = 0

    for line in lines:
        action, value = line.strip().split(" ")
        value = get_int(value)

        if action in action_funcs:
            aim += action_funcs[action](value)
        else:
            horizontal += value
            depth += (aim * value)

    print(depth * horizontal)

@add_argparse
def main(task):
    task_one() if task == "one" else task_two()


if __name__ == "__main__":
    main()

File: 02/cpp/main.cpp
#include <filesystem>
#include <fstream>
#include <string>
#include <iostream>
#include <vector>

#include <utils.h>

std::vector<std::string> split(std::string line) {
    std::vector<std::string> values;
    std::string temp;

    for (int i = 0; i < line.length(); i++) {
        if (line[i] == ' ') {
            values.push_back(temp);
            temp = "";
        }
        temp.push_back(line[i]);
    }

    values.push_back(temp);

    return values;
}


void one(){
    std::vector<std::string> values = textUtils::getFileLines("02/input_1.txt");

    int depth = 0;
    int horizontal = 0;

    for (std::string line: values) {
        std::vector<std::string> inputs = split(line);
        
        std::string action = inputs.at(0);
        int value = std::stoi(inputs.at(1));

        if (action == "up") {
            depth = depth - value;
        } else if (action == "down") {
            depth = depth + value;
        } else if (action == "forward") {
            horizontal = horizontal + value;
        }
    }

    std::cout << depth * horizontal << "\n";
}

void two(){
    std::vector<std::string> values = textUtils::getFileLines("02/input_1.txt");

    int depth = 0;
    int horizontal = 0;
    int aim = 0;

    for (std::string line: values) {
        std::vector<std::string> inputs = split(line);
        
        std::string action = inputs.at(0);
        int value = std::stoi(inputs.at(1));

        if (action == "up") {
            aim = aim - value;
        } else if (action == "down") {
            aim = aim + value;
        } else if (action == "forward") {
            horizontal = horizontal + value;
            depth = depth + (aim * value);
        }
    }

    std::cout << depth * horizontal << "\n";
}

int main() {
    one();
    two();
}